### PR TITLE
Create interface for Services and replace static methods with instance methods closes #110

### LINKED ISSUE #110: Create interface for Services and replace static methods with instance methods
SclService and SclAutomationService offer static methods.
The suggestion is to hide the implementation behind an interface and to use instance methods instead : 
- it allows client apps to mock the services when testing
- it allows different implementations for these services
- it better integrates with dependency injection frameworks

### PR DESCRIPTION
**Pull request by samirromdhani**
Fix https://github.com/com-pas/compas-sct/issues/110

This pull tries to rearrange the services in a more logical way and replace static methods with instance methods.

[creation] Interfaces implemented by existing services.
[update] UtilityClass is moved to the util package.

> BREAKING CHANGE: 💥 in sct-commons:

- The services which provide static methods are moved from scl to commons package.
Now  those services implements interfaces, no static methods.
Expect for PrivateService which is an UtilityClass is renamed to PrivateUtils and moved from scl to util package.

### CODE REVIEW

File: sct-commons/src/main/java/org/lfenergy/compas/sct/commons/service/impl/ExtRefService.java [hunk 1]
```diff
@@ -2,43 +2,37 @@
 //
 // SPDX-License-Identifier: Apache-2.0
 
-package org.lfenergy.compas.sct.commons.scl;
+package org.lfenergy.compas.sct.commons.service.impl;
```
Comment (gleizesDor) [inline sct-commons/src/main/java/org/lfenergy/compas/sct/commons/service/impl/ExtRefService.java:5]: don't use an "`impl`" package inside `service` but rather use an "`api`" package for interfaces inside `commons` package

File: sct-app/src/test/java/org.lfenergy.compas.sct.app/SclAutomationServiceIntegrationTest.java [hunk 2]
```diff
@@ -13,29 +13,35 @@
...
 import java.util.List;
 import java.util.UUID;
 
 import static org.assertj.core.api.Assertions.assertThat;
-import static org.assertj.core.api.Assertions.assertThatThrownBy;
 import static org.junit.jupiter.api.Assertions.*;
 import static org.lfenergy.compas.sct.commons.testhelpers.SclTestMarshaller.assertIsMarshallable;
 
-class SclAutomationServiceTest {
+class SclAutomationServiceIntegrationTest {
```
[the remark this answers is not in the document]
Comment (samirromdhani) [reply sct-app/src/test/java/org.lfenergy.compas.sct.app/SclAutomationServiceIntegrationTest.java:30]: Thanks for your comment. I thought so, but I used to not modify any dependency in this issue, that's why I did't used mockito and I kept it as integration test.
I will add it for you.

File: sct-commons/src/main/java/org/lfenergy/compas/sct/commons/dto/PrivateLinkedToSTDs.java [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+package org.lfenergy.compas.sct.commons.dto;
+
+import org.lfenergy.compas.scl2007b4.model.SCL;
+import org.lfenergy.compas.scl2007b4.model.TPrivate;
+
+import java.util.List;
+
+public record PrivateLinkedToSTDs(TPrivate tPrivate, List<SCL> stdList) {
```
[the remark this answers is not in the document]
Comment (samirromdhani) [reply sct-commons/src/main/java/org/lfenergy/compas/sct/commons/dto/PrivateLinkedToSTDs.java]: done

File: sct-app/src/main/java/org/lfenergy/compas/sct/app/SclAutomationService.java [hunk 4]
```diff
@@ -24,8 +25,12 @@
  *   <li>{@link SclAutomationService#createSCD(SCL, HeaderDTO, List) Adds all elements under the <b>SCL </b> object from given <b>SSD </b> and <b>STD </b> files}
  *  </ul>
  */
+@AllArgsConstructor
```
[the remark this answers is not in the document]
Comment (samirromdhani) [reply sct-app/src/main/java/org/lfenergy/compas/sct/app/SclAutomationService.java]: done

File: sct-app/src/test/java/org.lfenergy.compas.sct.app/SclAutomationServiceIntegrationTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (samirromdhani) [reply sct-app/src/test/java/org.lfenergy.compas.sct.app/SclAutomationServiceIntegrationTest.java:30]: New file added, [review [use mockito properly] ](https://github.com/com-pas/compas-sct/pull/318/commits/580a98a705e2e75352f3010f4eaba676cd180df1)

File: sct-commons/src/main/java/org/lfenergy/compas/sct/commons/service/ISubstationService.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+ *   <li>{@link ISubstationService#addSubstation(SCL, SCL) <em>Adds the <b>Substation </b> object from given <b>SCL </b> object</em>}</li>
+ * </ul>
+ * @see SubstationAdapter
+ * @see VoltageLevelAdapter
+ * @see org.lfenergy.compas.sct.commons.scl.sstation.BayAdapter
+ * @see org.lfenergy.compas.sct.commons.scl.sstation.FunctionAdapter
+ * @see org.lfenergy.compas.sct.commons.scl.sstation.LNodeAdapter
+ * @see PrivateUtils
+ */
+public interface ISubstationService {
```
[the remark this answers is not in the document]
Comment (samirromdhani) [reply sct-commons/src/main/java/org/lfenergy/compas/sct/commons/service/ISubstationService.java]: *Editor +1

File: sct-commons/src/main/java/org/lfenergy/compas/sct/commons/service/IExtRefService.java [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+import org.lfenergy.compas.scl2007b4.model.SCL;
+import org.lfenergy.compas.sct.commons.dto.ControlBlockNetworkSettings;
+import org.lfenergy.compas.sct.commons.dto.SclReportItem;
+import org.lfenergy.compas.sct.commons.util.ILDEPFSettings;
+import org.lfenergy.compas.sct.commons.util.Utils;
+
+import java.util.List;
+
+public interface IExtRefService {
```
[the remark this answers is not in the document]
Comment (samirromdhani) [reply sct-commons/src/main/java/org/lfenergy/compas/sct/commons/service/IExtRefService.java]: *Editor +1

File: sct-commons/src/main/java/org/lfenergy/compas/sct/commons/service/ISclService.java [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+ *
+ * </ul>
+ *
+ * @see HeaderAdapter
+ * @see SubstationAdapter
+ * @see IEDAdapter
+ * @see CommunicationAdapter
+ * @see DataTypeTemplateAdapter
+ */
+public interface ISclService {
```
[the remark this answers is not in the document]
Comment (samirromdhani) [reply sct-commons/src/main/java/org/lfenergy/compas/sct/commons/service/ISclService.java]: *Editor +1

File: sct-commons/src/main/java/org/lfenergy/compas/sct/commons/service/IHmiService.java [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+package org.lfenergy.compas.sct.commons.service;
+
+import org.lfenergy.compas.scl2007b4.model.SCL;
+import org.lfenergy.compas.scl2007b4.model.TFCDA;
+
+import java.util.List;
+
+public interface IHmiService {
```
[the remark this answers is not in the document]
Comment (samirromdhani) [reply sct-commons/src/main/java/org/lfenergy/compas/sct/commons/service/IHmiService.java]: *Editor +1

File: sct-commons/src/main/java/org/lfenergy/compas/sct/commons/service/impl/ExtRefService.java (hunk 1 above)
Comment (samirromdhani) [reply sct-commons/src/main/java/org/lfenergy/compas/sct/commons/service/impl/ExtRefService.java]: ok

File: sct-commons/src/main/java/org/lfenergy/compas/sct/commons/service/ISclService.java [hunk 9]
```diff
@@ -0,0 +1,308 @@
...
+    /**
+     * Initialise SCD file with Header and Private SCLFileType
+     *
+     * @param hId       SCL Header ID
+     * @param hVersion  SCL Header Version
+     * @param hRevision SCL Header Revision
+     * @return <em>SCL</em> SCD object
+     * @throws ScdException throws when inconsistance in SCL file
+     */
+    SCL initScl(final UUID hId, final String hVersion, final String hRevision) throws ScdException;
```
[the remark this answers is not in the document]
Comment (samirromdhani) [reply sct-commons/src/main/java/org/lfenergy/compas/sct/commons/service/ISclService.java]: Done.
I moved some methods from SclEditor to ExtrefEditor to correspond with their intended use.
Those methode are `updateExtRefBinders` and `updateExtRefSource`

File: sct-commons/src/test/java/org/lfenergy/compas/sct/commons/ExtRefServiceTest.java [hunk 10]
```diff
@@ -33,16 +40,20 @@
...
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
 import static org.lfenergy.compas.scl2007b4.model.TFCEnum.ST;
 import static org.lfenergy.compas.sct.commons.dto.ControlBlockNetworkSettings.*;
-import static org.lfenergy.compas.sct.commons.scl.ExtRefService.filterDuplicatedExtRefs;
 import static org.lfenergy.compas.sct.commons.testhelpers.SclHelper.*;
 import static org.lfenergy.compas.sct.commons.util.CommonConstants.*;
 import static org.lfenergy.compas.sct.commons.util.ControlBlockEnum.*;
 import static org.lfenergy.compas.sct.commons.util.SclConstructorHelper.newDurationInMilliSec;
 
+@ExtendWith(MockitoExtension.class)
```
[the remark this answers is not in the document]
Comment (samirromdhani) [reply sct-commons/src/test/java/org/lfenergy/compas/sct/commons/ExtRefServiceTest.java:51]: We can use Mockito even if we don't have any dependencies to mock.
As stated in many articles and tutorials .
> @InjectMocks annotation which tries to do constructor, method or field dependency injection of mock objects in to other type. It does not try to be a full features dependency injection framework but is helpful to wire up objects to be tested.

In addition this is about conventions and making development experience easier. That means that the we do not have to spend a lot of time configuring files in order to get setup.
See https://en.wikipedia.org/wiki/Convention_over_configuration.

**Review by massifben — APPROVED** [review]

File: sct-commons/src/test/java/org/lfenergy/compas/sct/commons/ExtRefServiceTest.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (samirromdhani) [reply sct-commons/src/test/java/org/lfenergy/compas/sct/commons/ExtRefServiceTest.java:51]: As we discussed, it's nice-to-have until we have a solid case to support the no use of Mockito when we don't have dependencies to mock.

File: sct-commons/src/test/java/org/lfenergy/compas/sct/commons/SubstationServiceTest.java [hunk 11]
```diff
@@ -2,29 +2,36 @@
...
 import org.lfenergy.compas.sct.commons.testhelpers.SclTestMarshaller;
+import org.mockito.InjectMocks;
+import org.mockito.junit.jupiter.MockitoExtension;
 
 import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
-import static org.junit.jupiter.api.Assertions.*;
+import static org.junit.jupiter.api.Assertions.assertThrows;
 import static org.lfenergy.compas.sct.commons.testhelpers.SclTestMarshaller.assertIsMarshallable;
 
+@ExtendWith(MockitoExtension.class)
```
[the remark this answers is not in the document]
Comment (samirromdhani) [reply sct-commons/src/test/java/org/lfenergy/compas/sct/commons/SubstationServiceTest.java:21]: https://github.com/com-pas/compas-sct/pull/318#discussion_r1308687376

File: sct-commons/src/test/java/org/lfenergy/compas/sct/commons/SclElementsProviderServiceTest.java [hunk 12]
```diff
@@ -0,0 +1,343 @@
...
+import java.util.*;
+import java.util.stream.Collectors;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.junit.jupiter.api.Assertions.*;
+import static org.lfenergy.compas.sct.commons.testhelpers.DataTypeUtils.createDa;
+import static org.lfenergy.compas.sct.commons.testhelpers.DataTypeUtils.createDo;
+
+@ExtendWith(MockitoExtension.class)
```
[the remark this answers is not in the document]
Comment (samirromdhani) [reply sct-commons/src/test/java/org/lfenergy/compas/sct/commons/SclElementsProviderServiceTest.java:28]: https://github.com/com-pas/compas-sct/pull/318#discussion_r1308687376

File: sct-commons/src/test/java/org/lfenergy/compas/sct/commons/SclEditorServiceTest.java [hunk 13]
```diff
@@ -2,37 +2,41 @@
...
 import static org.junit.jupiter.api.Assertions.*;
-import static org.lfenergy.compas.sct.commons.testhelpers.DataTypeUtils.createDa;
-import static org.lfenergy.compas.sct.commons.testhelpers.DataTypeUtils.createDo;
 import static org.lfenergy.compas.sct.commons.testhelpers.SclHelper.LD_SUIED;
 import static org.lfenergy.compas.sct.commons.testhelpers.SclHelper.getDAIAdapters;
 import static org.lfenergy.compas.sct.commons.testhelpers.SclTestMarshaller.assertIsMarshallable;
 import static org.lfenergy.compas.sct.commons.util.PrivateEnum.COMPAS_SCL_FILE_TYPE;
 
-class SclServiceTest {
+@ExtendWith(MockitoExtension.class)
```
[the remark this answers is not in the document]
Comment (samirromdhani) [reply sct-commons/src/test/java/org/lfenergy/compas/sct/commons/SclEditorServiceTest.java:35]: https://github.com/com-pas/compas-sct/pull/318#discussion_r1308687376

File: sct-commons/src/test/java/org/lfenergy/compas/sct/commons/HmiServiceTest.java [hunk 14]
```diff
@@ -2,32 +2,39 @@
...
 
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.lfenergy.compas.sct.commons.testhelpers.SclHelper.*;
 import static org.lfenergy.compas.sct.commons.util.SclConstructorHelper.newFcda;
 
+@ExtendWith(MockitoExtension.class)
 class HmiServiceTest {
 
+    @InjectMocks
+    HmiService hmiService;
```
[the remark this answers is not in the document]
Comment (samirromdhani) [reply sct-commons/src/test/java/org/lfenergy/compas/sct/commons/HmiServiceTest.java:29]: https://github.com/com-pas/compas-sct/pull/318#discussion_r1308687376

File: sct-commons/src/test/java/org/lfenergy/compas/sct/commons/ExtRefServiceTest.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (samirromdhani) [reply sct-commons/src/test/java/org/lfenergy/compas/sct/commons/ExtRefServiceTest.java:51]: [Examples of correct usage of @InjectMocks:](https://github.com/mockito/mockito/blob/2164df10daba44acf18bcd5b8886f3b4addf7b24/src/main/java/org/mockito/internal/exceptions/Reporter.java#L826)

**Pull request merged**